File: resources/js/Utils/recommender.js
/*
 * Rekomendasi algoritma dari karakteristik dataset.
 *
 * Padanan `python/core/recommender.py`. Alurnya: profil kolom -> karakteristik
 * -> daftar algoritma yang cocok beserta alasannya. Alasan selalu menyebut
 * kolom yang mendasarinya, supaya rekomendasi bisa ditelusuri pengguna dan tidak
 * terbaca sebagai tebakan.
 *
 * Rekomendasi tidak pernah membatasi pilihan — seluruh algoritma tetap bisa
 * dijalankan, yang direkomendasikan hanya ditandai dan diurutkan lebih dulu.
 */

// Kolom kategori dengan kardinalitas di atas ini lebih mirip identitas
// (kode pos, SKU) daripada label yang bisa diprediksi.
const MAX_TARGET_CLASSES = 8;

// Ambang lift untuk menyebut dua item "sering muncul bersama". Di bawah ini
// pasangannya tidak lebih sering daripada yang diharapkan secara kebetulan.
const BASKET_LIFT = 1.4;
const BASKET_SUPPORT = 0.05;

/*
 * Seberapa kuat nilai antar kolom kategori muncul bersama.
 *
 * Tanpa ukuran ini, "data transaksional" hanya berarti "punya kolom identitas
 * dan beberapa kolom kategori" — dan itu benar untuk hampir semua tabel,
 * sehingga Association Rule akan direkomendasikan di mana-mana. Yang menentukan
 * sebenarnya adalah ada tidaknya pasangan nilai yang benar-benar berulang
 * bersamaan.
 */
function basketStrength(table, columns) {
    if (columns.length < 2 || !table?.rows?.length) {
        return 0;
    }

    const total = table.rows.length;
    const singles = new Map();
    const pairs = new Map();

    for (const row of table.rows) {
        const items = columns
            .filter((column) => row[column] !== null && row[column] !== undefined)
            .map((column) => ({ column, value: `${column}=${row[column]}` }));

        for (const item of items) {
            singles.set(item.value, (singles.get(item.value) ?? 0) + 1);
        }

        for (let i = 0; i < items.length; i += 1) {
            for (let j = i + 1; j < items.length; j += 1) {
                const key = `${items[i].value}|${items[j].value}`;

                pairs.set(key, (pairs.get(key) ?? 0) + 1);
            }
        }
    }

    let highest = 0;

    for (const [key, count] of pairs.entries()) {
        const support = count / total;

        if (support < BASKET_SUPPORT) {
            continue;
        }

        const [left, right] = key.split('|');
        const expected =
            ((singles.get(left) ?? 0) / total) * ((singles.get(right) ?? 0) / total);
        const lift = expected ? support / expected : 0;

        highest = Math.max(highest, lift);
    }

    return highest;
}

/** Kandidat target: kategori berkardinalitas rendah dan numerik non-identitas. */
export function analyzeCharacteristics(profile, table) {
    const categoricalTargets = profile.categorical
        .filter(
            (column) => column.unique >= 2 && column.unique <= MAX_TARGET_CLASSES,
        )
        .sort((a, b) => a.unique - b.unique);

    // Kolom numerik yang berhubungan dengan kolom numerik lain lebih layak jadi
    // target regresi daripada kolom yang berdiri sendiri.
    const numericTargets = profile.numeric
        .map((column) => {
            const index = profile.correlation?.columns.indexOf(column.name) ?? -1;
            const strongest =
                index >= 0
                    ? Math.max(
                          ...profile.correlation.matrix[index].map((value, other) =>
                              other === index ? 0 : Math.abs(value),
                          ),
                      )
                    : 0;

            return { column, strongest };
        })
        .sort((a, b) => b.strongest - a.strongest);

    const basketColumns = profile.categorical.filter(
        (column) => column.unique >= 2 && column.unique <= 12,
    );

    const basketLift = basketStrength(
        table,
        basketColumns.map((column) => column.name),
    );

    return {
        basketLift,
        rowCount: profile.rowCount,
        numericCount: profile.numeric.length,
        categoricalCount: profile.categorical.length,
        datetimeCount: profile.datetime.length,
        hasIdentifier: profile.identifiers.length > 0,
        outlierRatio: profile.outlierRatio,
        duplicateRows: profile.duplicateRows,
        categoricalTargets,
        numericTargets,
        basketColumns,
        // Dataset transaksional: ada nomor identitas per baris, minimal dua
        // kolom "item", DAN pasangan nilainya memang berulang bersamaan.
        isTransactional:
            profile.identifiers.length > 0 &&
            basketColumns.length >= 2 &&
            basketLift >= BASKET_LIFT,
    };
}

const percent = (value) => `${(value * 100).toFixed(1).replace('.', ',')}%`;

/**
 * @returns {Array<{ key, level: 'high'|'medium', reason }>} urut dari yang
 *   paling sesuai.
 */
export function recommendAlgorithms(profile, table) {
    const facts = analyzeCharacteristics(profile, table);
    const recommendations = [];

    const categoricalTarget = facts.categoricalTargets[0];
    const numericTarget = facts.numericTargets[0];

    if (categoricalTarget) {
        recommendations.push({
            key: 'classification',
            level: 'high',
            reason: `Kolom "${categoricalTarget.name}" berisi ${categoricalTarget.unique} kelas — cocok dijadikan target klasifikasi.`,
        });
    }

    if (numericTarget && numericTarget.strongest >= 0.4) {
        recommendations.push({
            key: 'regression',
            level: 'high',
            reason: `Kolom "${numericTarget.column.name}" berkorelasi ${numericTarget.strongest
                .toFixed(2)
                .replace('.', ',')} dengan kolom numerik lain — nilainya dapat diprediksi.`,
        });
    } else if (numericTarget) {
        recommendations.push({
            key: 'regression',
            level: 'medium',
            reason: `Ada ${facts.numericCount} kolom numerik, tetapi korelasi antar kolom masih lemah.`,
        });
    }

    if (facts.datetimeCount > 0 && facts.numericCount > 0) {
        recommendations.push({
            key: 'timeseries',
            level: 'high',
            reason: `Kolom waktu "${profile.datetime[0].name}" tersedia, sehingga nilai numerik bisa dianalisis sebagai deret waktu.`,
        });
    }

    if (facts.isTransactional) {
        recommendations.push({
            key: 'association',
            level: 'high',
            reason: `Berbentuk transaksi: ${facts.basketColumns.length} kolom item, dengan pasangan nilai yang muncul bersama ${facts.basketLift
                .toFixed(1)
                .replace('.', ',')}× lebih sering daripada kebetulan.`,
        });
    }

    if (facts.outlierRatio >= 0.02) {
        recommendations.push({
            key: 'anomaly',
            level: 'high',
            reason: `${percent(facts.outlierRatio)} nilai numerik berada di luar batas wajar — jumlah nilai ekstremnya menonjol.`,
        });
    } else {
        recommendations.push({
            key: 'anomaly',
            level: 'medium',
            reason: `Nilai ekstrem hanya ${percent(facts.outlierRatio)} dari data, anomali kemungkinan sedikit.`,
        });
    }

    if (!categoricalTarget) {
        recommendations.push({
            key: 'clustering',
            level: 'high',
            reason: 'Tidak ada kolom target yang jelas — pengelompokan tanpa label paling masuk akal.',
        });
    } else if (facts.numericCount >= 2) {
        recommendations.push({
            key: 'clustering',
            level: 'medium',
            reason: `Ada ${facts.numericCount} kolom numerik yang bisa dikelompokkan sebagai segmentasi tambahan.`,
        });
    }

    const order = { high: 0, medium: 1 };

    return recommendations.sort((a, b) => order[a.level] - order[b.level]);
}

File: python/core/recommender.py
"""Auto Recommendation: menyarankan analisis yang sesuai dengan dataset.

Setiap saran wajib menyebutkan kolom yang mendasarinya. Rekomendasi tanpa alasan
tidak bisa ditelusuri pengguna dan akan terbaca sebagai tebakan; dengan alasan,
pengguna bisa menilai sendiri apakah dugaan sistem tentang datanya benar.

Rekomendasi tidak pernah mengunci pilihan — seluruh analisis tetap dapat
dijalankan, yang disarankan hanya ditandai dan diurutkan lebih dulu.
"""

from __future__ import annotations

import pandas as pd

from core import loader

MAX_TARGET_CLASSES = 12
STRONG_CORRELATION = 0.4
OUTLIER_ALERT = 0.02
BASKET_LIFT = 1.4
BASKET_SUPPORT = 0.05


def characteristics(frame: pd.DataFrame) -> dict:
    numeric = loader.numeric_columns(frame)
    categorical = loader.categorical_columns(frame)
    datetimes = loader.datetime_columns(frame)
    identifiers = [name for name in frame.columns if loader.is_identifier(frame[name])]

    target_candidates = [
        name for name in categorical
        if 2 <= frame[name].nunique(dropna=True) <= MAX_TARGET_CLASSES
    ]

    # Kolom numerik yang berhubungan dengan kolom numerik lain lebih layak jadi
    # target regresi daripada kolom yang berdiri sendiri.
    strongest = {}

    if len(numeric) >= 2:
        matrix = frame[numeric].corr(numeric_only=True).abs()

        for name in numeric:
            others = matrix[name].drop(labels=[name], errors="ignore")
            strongest[name] = float(others.max()) if not others.empty else 0.0

    basket = [name for name in categorical if frame[name].nunique(dropna=True) <= 12]

    return {
        "row_count": int(len(frame)),
        "numeric": numeric,
        "categorical": categorical,
        "datetime": datetimes,
        "identifiers": identifiers,
        "target_candidates": sorted(
            target_candidates, key=lambda name: frame[name].nunique()
        ),
        "numeric_strength": strongest,
        "basket_columns": basket,
        "basket_lift": _basket_strength(frame, basket),
        "outlier_ratio": _outlier_ratio(frame, numeric),
    }


def _outlier_ratio(frame: pd.DataFrame, numeric: list) -> float:
    if not numeric or frame.empty:
        return 0.0

    total = 0

    for name in numeric:
        values = frame[name].dropna()

        if values.empty:
            continue

        q1, q3 = values.quantile(0.25), values.quantile(0.75)
        iqr = q3 - q1
        total += int(((values < q1 - 1.5 * iqr) | (values > q3 + 1.5 * iqr)).sum())

    return round(total / (len(numeric) * len(frame)), 6)


def _basket_strength(frame: pd.DataFrame, columns: list) -> float:
    """Seberapa kuat nilai antar kolom kategori muncul bersamaan.

    Tanpa ukuran ini, "data transaksional" hanya berarti "punya beberapa kolom
    kategori" — dan itu benar untuk hampir semua tabel, sehingga Association Rule
    akan disarankan di mana-mana. Yang menentukan sebenarnya adalah ada tidaknya
    pasangan nilai yang benar-benar berulang bersamaan.
    """
    if len(columns) < 2 or frame.empty:
        return 0.0

    total = len(frame)
    highest = 0.0

    for i, left in enumerate(columns):
        for right in columns[i + 1:]:
            pair = frame[[left, right]].dropna()

            if pair.empty:
                continue

            joint = pair.groupby([left, right], observed=True).size() / total
            left_share = pair[left].value_counts(normalize=True)
            right_share = pair[right].value_counts(normalize=True)

            for (left_value, right_value), support in joint.items():
                if support < BASKET_SUPPORT:
                    continue

                expected = left_share.get(left_value, 0) * right_share.get(right_value, 0)

                if expected > 0:
                    highest = max(highest, support / expected)

    return round(float(highest), 3)


def run(params: dict) -> dict:
    frame = loader.load(params)
    facts = characteristics(frame)
    suggestions = []

    targets = facts["target_candidates"]
    numeric_strength = facts["numeric_strength"]

    if targets:
        target = targets[0]
        classes = int(frame[target].nunique())
        suggestions.append({
            "key": "classification",
            "name": "Classification",
            "level": "high",
            "reason": f'Kolom "{target}" berisi {classes} kelas — cocok dijadikan target klasifikasi.',
            "suggested_target": target,
        })

    if numeric_strength:
        best = max(numeric_strength, key=numeric_strength.get)
        strength = numeric_strength[best]

        suggestions.append({
            "key": "regression",
            "name": "Regression",
            "level": "high" if strength >= STRONG_CORRELATION else "medium",
            "reason": (
                f'Kolom "{best}" berkorelasi {strength:.2f} dengan kolom numerik lain — '
                "nilainya dapat diprediksi."
                if strength >= STRONG_CORRELATION
                else f"Ada {len(facts['numeric'])} kolom numerik, tetapi korelasi antar kolom masih lemah."
            ),
            "suggested_target": best,
        })

    if facts["datetime"] and facts["numeric"]:
        suggestions.append({
            "key": "forecasting",
            "name": "Forecasting",
            "level": "high",
            "reason": (
                f'Kolom waktu "{facts["datetime"][0]}" tersedia, sehingga nilai numerik '
                "bisa dianalisis dan diproyeksikan sebagai deret waktu."
            ),
            "suggested_target": facts["numeric"][0],
        })

    if facts["identifiers"] and len(facts["basket_columns"]) >= 2 and facts["basket_lift"] >= BASKET_LIFT:
        suggestions.append({
            "key": "association",
            "name": "Association Rule",
            "level": "high",
            "reason": (
                f"Berbentuk transaksi: {len(facts['basket_columns'])} kolom item, dengan "
                f"pasangan nilai yang muncul bersama {facts['basket_lift']:.1f}× lebih sering "
                "daripada kebetulan."
            ),
        })

    ratio = facts["outlier_ratio"]
    suggestions.append({
        "key": "anomaly",
        "name": "Anomaly Detection",
        "level": "high" if ratio >= OUTLIER_ALERT else "medium",
        "reason": (
            f"{ratio * 100:.1f}% nilai numerik berada di luar batas wajar — "
            "jumlah nilai ekstremnya menonjol."
            if ratio >= OUTLIER_ALERT
            else f"Nilai ekstrem hanya {ratio * 100:.1f}% dari data, anomali kemungkinan sedikit."
        ),
    })

    if not targets:
        suggestions.append({
            "key": "clustering",
            "name": "Clustering",
            "level": "high",
            "reason": "Tidak ada kolom target yang jelas — pengelompokan tanpa label paling masuk akal.",
        })
    elif len(facts["numeric"]) >= 2:
        suggestions.append({
            "key": "clustering",
            "name": "Clustering",
            "level": "medium",
            "reason": (
                f"Ada {len(facts['numeric'])} kolom numerik yang bisa dikelompokkan "
                "sebagai segmentasi tambahan."
            ),
        })

    order = {"high": 0, "medium": 1}
    suggestions.sort(key=lambda item: order.get(item["level"], 2))

    return {
        "characteristics": {
            "row_count": facts["row_count"],
            "numeric_count": len(facts["numeric"]),
            "categorical_count": len(facts["categorical"]),
            "datetime_count": len(facts["datetime"]),
            "identifier_count": len(facts["identifiers"]),
            "outlier_ratio": facts["outlier_ratio"],
            "basket_lift": facts["basket_lift"],
            "target_candidates": targets[:5],
            "is_transactional": bool(
                facts["identifiers"]
                and len(facts["basket_columns"]) >= 2
                and facts["basket_lift"] >= BASKET_LIFT
            ),
        },
        "recommendations": suggestions,
    }
